### PR TITLE
fix: close sidebar instead of closing window when opening conversation [SQSERVICES-1882]

### PR DESCRIPTION
**Pull request by V-Gira**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/SQSERVICES-1882" title="SQSERVICES-1882" target="_blank">BugSQSERVICES-1882</a>  When coming from team settings, webapp crashes when opening a new group. 
  </td></table>
  <br />
 

[SQSERVICES-1882]

### Issue
- If a user comes from the team settings panel into the webapp (immediate-login), the webapp (the whole tab) will close when joining a group via search.
Peek 2023-01-30 13-21

### Solution
- replace `close` by `onClose` in the `openConversation` promise

#### What happened?
- Here's a quote from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/Window/close) about the `Window.close()` method: 
> `The Window.close()` method closes the current window, or the window on which it was called.
This method can only be called on windows that were opened by a script using the [Window.open()](https://developer.mozilla.org/en-US/docs/Web/API/Window/open) method.

- Opening the app from Team Setting's sidebar uses the `Window.open()` method
- The `Window.close()` method in `openConversation` would only result in an error if the app was opened normally, and close the tab if the app was opened through Team Setting (or any external link using the `Window.open()` method)

[SQSERVICES-1882]: https://wearezeta.atlassian.net/browse/SQSERVICES-1882?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx [hunk 1]
```diff
@@ -138,7 +138,7 @@ const StartUI: React.FC<StartUIProps> = ({
   const openInviteModal = () => showInviteModal({userState});
 
   const openConversation = (conversation: Conversation): Promise<void> => {
-    return actions.openGroupConversation(conversation).then(close);
+    return actions.openGroupConversation(conversation).then(onClose);
```
Comment (V-Gira) [inline src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx]: this alone solves the issue, the rest of the PR is some small refactoring

**Review by PatrykBuniX — APPROVED** [review]
Nice catch

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx [hunk 2]
```diff
@@ -100,27 +100,27 @@ const StartUI: React.FC<StartUIProps> = ({
...
   const peopleSearchResults = useRef<SearchResultsData | undefined>(undefined);
 
-  const openFirstConversation = (): void => {
+  const openFirstConversation = async (): Promise<void> => {
     if (peopleSearchResults.current) {
       const {contacts, groups} = peopleSearchResults.current;
       if (contacts.length > 0) {
-        openContact(contacts[0]);
+        await openContact(contacts[0]);
         return;
```
Comment (PatrykBuniX) [inline src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx]: you can just return an async function here -> `return openContact(contacts[0]);` (same thing with the ones below)

**Review by przemvs — APPROVED** [review]

File: src/script/page/LeftSidebar/panels/StartUI/components/groupList/GroupListItem.tsx [hunk 3]
```diff
@@ -43,8 +43,9 @@ const GroupListItem: React.FC<GroupListItemProps> = ({click, group}) => {
   } = useKoSubscribableChildren(group, ['display_name', 'participating_user_ets', 'is1to1']);
 
   const onClick = () => {
+    const url = generateConversationUrl(group.qualifiedId);
     click(group);
-    navigate(generateConversationUrl(group.qualifiedId));
+    createNavigate(url);
```
Comment (atomrc) [inline src/script/page/LeftSidebar/panels/StartUI/components/groupList/GroupListItem.tsx]: This change seems suspicious to me. You create something but never return it, this function is never consumed 🤔

File: src/script/view_model/ActionsViewModel.ts [hunk 4]
```diff
@@ -110,8 +110,8 @@ export class ActionsViewModel {
     return new Promise(resolve => {
       PrimaryModal.show(PrimaryModal.type.CONFIRM, {
         primaryAction: {
-          action: () => {
-            this.connectionRepository.cancelRequest(userEntity, hideConversation, nextConversationEntity);
+          action: async () => {
+            return this.connectionRepository.cancelRequest(userEntity, hideConversation, nextConversationEntity);
             resolve();
```
Comment (atomrc) [inline src/script/view_model/ActionsViewModel.ts]: This is wrong, the `resolve` will never be called

File: src/script/view_model/ActionsViewModel.ts [hunk 5]
```diff
@@ -203,8 +203,8 @@ export class ActionsViewModel {
       return new Promise(resolve => {
         PrimaryModal.show(PrimaryModal.type.CONFIRM, {
           primaryAction: {
-            action: () => {
-              this.messageRepository.deleteMessage(conversationEntity, messageEntity);
+            action: async () => {
+              return this.messageRepository.deleteMessage(conversationEntity, messageEntity);
               resolve();
```
Comment (atomrc) [inline src/script/view_model/ActionsViewModel.ts]: Same here, the `resolve` will not be called

**Review by atomrc — CHANGES_REQUESTED** [review]
A few floating promises have been fixed... The wrong way :)

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx [hunk 6]
```diff
@@ -100,28 +100,26 @@ const StartUI: React.FC<StartUIProps> = ({
 
   const peopleSearchResults = useRef<SearchResultsData | undefined>(undefined);
 
-  const openFirstConversation = (): void => {
+  const openFirstConversation = async (): Promise<void> => {
```
Comment (atomrc) [inline src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx:103]: you actually don't need to mark this one async. 

The rule of thumb is: you only need to put `async` if you need `await` in the body of the function. 

Else a simple `: Promise<...>` signature is enough :) 

```suggestion
  const openFirstConversation = (): Promise<void> => {
```

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx [hunk 7]
```diff
@@ -100,28 +100,26 @@ const StartUI: React.FC<StartUIProps> = ({
...
   };
 
   const openContact = async (user: User) => {
     const conversationEntity = await actions.getOrCreate1to1Conversation(user);
-    actions.open1to1Conversation(conversationEntity);
+    return actions.open1to1Conversation(conversationEntity);
   };
 
-  const openOther = (user: User) => {
+  const openOther = async (user: User) => {
```
Comment (atomrc) [inline src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx]: see https://github.com/wireapp/wire-webapp/pull/14586/files#r1090618504

File: src/script/view_model/ActionsViewModel.ts [hunk 8]
```diff
@@ -330,9 +330,9 @@ export class ActionsViewModel {
     return this.openConversation(conversationEntity);
   };
 
-  private readonly openConversation = (conversationEntity: Conversation): void => {
+  private readonly openConversation = async (conversationEntity: Conversation): Promise<void> => {
     if (conversationEntity.is_archived()) {
-      this.conversationRepository.unarchiveConversation(conversationEntity, true);
+      return this.conversationRepository.unarchiveConversation(conversationEntity, true);
```
Comment (atomrc) [inline src/script/view_model/ActionsViewModel.ts]: This changes the behavior (the amplify event will not be fired)

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx (hunk 6 above)
Comment (V-Gira) [reply src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx:103]: I think I'm missing something obvious :thinking: 
image

**Review by atomrc — CHANGES_REQUESTED** [review]

File: src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx (hunk 6 above)
Comment (atomrc) [reply src/script/page/LeftSidebar/panels/StartUI/StartUI.tsx:103]: I guess you can just remove the return type at all, it's just extra maintenance here :) 
TypeScript is going to infer it for you ;) 

```suggestion
  const openFirstConversation = () => {
```

**Review by atomrc — APPROVED** [review]

**Pull request merged**